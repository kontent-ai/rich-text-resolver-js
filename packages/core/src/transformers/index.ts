export * from "./html-transformer/index.js";
export * from "./portable-text-transformer/index.js";
export * from "./transformer-models.js";
