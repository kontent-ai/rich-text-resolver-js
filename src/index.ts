export { toVueImageDefault } from './utils/resolution/vue.js';
export { toHTMLImageDefault } from './utils/resolution/html.js';
export { traversePortableText } from './utils/transformer-utils.js';
export { transformToJson } from "./transformers/json-transformer/json-transformer.js";
export { transformToPortableText } from "./transformers/portable-text-transformer/portable-text-transformer.js";
export { parse as browserParse } from "./parser/browser/rich-text-browser-parser.js";
export { parse as nodeParse } from "./parser/node/rich-text-node-parser.js";
