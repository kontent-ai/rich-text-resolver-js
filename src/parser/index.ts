import * as runtimeEnvironment from "browser-or-node";
import { browserParse } from "./browser/index.js";
import { nodeParse } from "./node/index.js";

export * from "./parser-models.js";

export const parseHTML = (htmlInput: string) =>
  runtimeEnvironment.isBrowser
    ? browserParse(htmlInput)
    : nodeParse(htmlInput);
