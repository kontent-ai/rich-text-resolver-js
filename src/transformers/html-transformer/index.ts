export * from "./html-transformer.js";
