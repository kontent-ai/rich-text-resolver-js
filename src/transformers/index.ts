export * from "./json-transformer/index.js";
export * from "./portable-text-transformer/index.js";
export * from "./transformer-models.js";
