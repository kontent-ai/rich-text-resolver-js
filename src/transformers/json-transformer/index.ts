export * from './json-transformer.js';