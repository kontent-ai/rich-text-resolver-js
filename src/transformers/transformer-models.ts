import {
  ArbitraryTypedObject,
  PortableTextBlock,
  PortableTextListItemBlock,
  PortableTextMarkDefinition,
  PortableTextSpan,
} from "@portabletext/types";

import {
  blockElements,
  ignoredElements,
  listTypeElements,
  markElements,
  textStyleElements,
  validElements,
} from "../utils/constants.js";

type LiteralOrString<T extends string> = T | (string & {});

/**
 * A reference to various Kontent.ai objects in rich text
 */
export interface Reference extends ArbitraryTypedObject {
  _type: "reference";
  /**
   * An identifier of the referenced object
   */
  _ref: string;
  /**
   * Type of reference (codename, id or external id)
   */
  referenceType: "codename" | "external-id" | "id";
}

/**
 * Represents an asset object used in rich text.
 */
export interface AssetReference extends Reference {
  /**
   * URL of an asset.
   */
  url: string;
  /**
   * Alternate image text.
   */
  alt?: string;
}

/**
 * Represents a mark definition for a link to an external URL in rich text element. This includes asset, phone and email links.
 */
export interface PortableTextExternalLink extends PortableTextMarkDefinition {
  _type: "link";
  href?: string;
  rel?: string;
  title?: string;
}

/**
 * Represents a mark definition for a link to a content item in rich text element.
 */
export interface PortableTextItemLink extends PortableTextMarkDefinition {
  _type: "contentItemLink";
  reference: Reference;
}

/**
 * Represents an inline image used in rich text element.
 */
export interface PortableTextImage extends ArbitraryTypedObject {
  _type: "image";
  /**
   * Reference to an asset used in rich text.
   */
  asset: AssetReference;
}

/**
 * Represents a table in rich text element.
 */
export interface PortableTextTable extends ArbitraryTypedObject {
  _type: "table";
  /**
   * Array of table row objects.
   */
  rows: PortableTextTableRow[];
}

/**
 * Represents a single row of cells in a portable text table.
 */
export interface PortableTextTableRow extends ArbitraryTypedObject {
  _type: "row";
  /**
   * Array of table cells.
   */
  cells: PortableTextTableCell[];
}

/**
 * Represents a single cell in a portable text table.
 */
export interface PortableTextTableCell extends ArbitraryTypedObject {
  _type: "cell";
  /**
   * All blocks belonging to a cell.
   */
  content: PortableTextObject[];
}

/**
 * Represents a component or a linked item used in rich text.
 */
export interface PortableTextComponentOrItem extends ArbitraryTypedObject {
  _type: "componentOrItem";
  /**
   * `component` for components or `item | link` for linked items
   */
  dataType: ModularContentType;
  /**
   * Reference to a component or a linked item.
   */
  component: Reference;
}

/**
 * Represents either a style (strong, em, etc.) or references a link object in markDefs array
 */
export interface PortableTextMark extends ArbitraryTypedObject {
  _type: "mark";
  value: LiteralOrString<TextStyleElement>; // value can be a shortguid (string) if a mark references a link
}

/**
 * Represents a block, usually a paragraph or heading.
 *
 * Narrows the `_type` to `block` for type guard purposes.
 */
export interface PortableTextStrictBlock extends Omit<PortableTextBlock, "_type">, ArbitraryTypedObject {
  _type: "block";
}

/**
 * Represents a list item block. Similar to regular block but requires `listItem` property.
 *
 * Narrows the `_type` to `block` for type guard purposes.
 */
export interface PortableTextStrictListItemBlock
  extends Omit<PortableTextListItemBlock, "_type">, ArbitraryTypedObject {
  _type: "block";
}

export type PortableTextLink =
  | PortableTextItemLink
  | PortableTextExternalLink;

/**
 * Union of all default, top-level portable text object types.
 */
export type PortableTextObject =
  | PortableTextComponentOrItem
  | PortableTextImage
  | PortableTextTable
  | PortableTextStrictBlock
  | PortableTextStrictListItemBlock;

/**
 * Union of all nested portable text object types.
 */
export type PortableTextInternalObject =
  | Reference
  | PortableTextMark
  | PortableTextLink
  | PortableTextTableRow
  | PortableTextTableCell
  | PortableTextSpan;

/**
 * Union of all default portable text object types.
 */
export type PortableTextItem = PortableTextObject | PortableTextInternalObject;

/**
 * `link` represent a rich text linked item in delivery API context
 */
type DeliveryLinkedItem = "link";

/**
 * `item` represents a rich text linked item in management API context
 */
type ManagementLinkedItem = "item";

/**
 * Represents type of modular content (items, components) in different rich text contexts
 */
export type ModularContentType = "component" | DeliveryLinkedItem | ManagementLinkedItem;

/**
 * Re-exports all types from the package, to allow both custom types and
 * predefined types to be imported from a single point via "exports" in package.json.
 */
export * from "@portabletext/types";

export type TextStyleElement = typeof textStyleElements[number];
export type BlockElement = typeof blockElements[number];
export type IgnoredElement = typeof ignoredElements[number];
export type MarkElement = typeof markElements[number];
export type ValidElement = typeof validElements[number];
export type ListTypeElement = typeof listTypeElements[number];
export type ShortGuid = string;
