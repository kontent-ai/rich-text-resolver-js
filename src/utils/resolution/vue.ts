import {
  PortableTextImage,
  PortableTextObject,
  PortableTextTable,
  PortableTextTableCell,
  PortableTextTableRow,
  toHTML,
} from "../../../index.js";

/**
 * Renders a portable text table to a Vue virtual DOM node.
 *
 * @param {PortableTextTable} table - The portable text table object to render.
 * @param {(value: PortableTextBlock[]) => string} resolver - A function that resolves
 *        the content of each cell in the table.
 * @param {Function} vueRenderFunction - A Vue render function, typically the `h` function from Vue.
 * @returns {VueNode} The rendered table as a Vue virtual DOM node.
 */
export const resolveTable = (
  table: PortableTextTable,
  vueRenderFunction: Function,
  resolver: (value: PortableTextObject[]) => string = toHTML,
) => {
  const renderCell = (cell: PortableTextTableCell) => {
    const cellContent = resolver(cell.content);
    return vueRenderFunction("td", {}, cellContent);
  };

  const renderRow = (row: PortableTextTableRow) => {
    const cells = row.cells.map(renderCell);
    return vueRenderFunction("tr", {}, cells);
  };

  const renderRows = () => table.rows.map(renderRow);

  return vueRenderFunction("table", {}, renderRows());
};

/**
 * Resolves an image object to a Vue virtual DOM node using a provided Vue render function.
 *
 * @param {PortableTextImage} image - The portable text image object to be rendered.
 * @param {Function} vueRenderFunction - A Vue render function, typically the `h` function from Vue.
 * @param {(image: PortableTextImage) => VueImage} resolver - A function that takes an image object
 *        and returns an object with `src` and `alt` properties, and possibly other HTML attributes.
 *        Default implementation provided if not specified.
 * @returns {VueNode} The resolved image as a Vue virtual DOM node.
 */
export const resolveImage = (
  image: PortableTextImage,
  vueRenderFunction: Function,
  resolver: (image: PortableTextImage) => VueImage = toVueImageDefault,
) => vueRenderFunction("img", resolver(image));

/**
 * Provides a default resolver function for an image object to Vue. Default fallback for `resolver`
 * argument of `resolveImage` function.
 *
 * @param {PortableTextImage} image - The portable text image object to be rendered.
 * @returns {VueImage} An object representing the image, containing `src` and `alt` properties,
 *          and potentially other HTML attributes.
 */
export const toVueImageDefault = (image: PortableTextImage): VueImage => ({
  src: image.asset.url,
  alt: image.asset.alt || "",
});

type VueImage = { src: string; alt: string; [key: string]: any };
