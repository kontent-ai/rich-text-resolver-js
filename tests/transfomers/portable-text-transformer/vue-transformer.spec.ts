import { Elements,ElementType } from "@kontent-ai/delivery-sdk";
import { PortableText, PortableTextComponentProps, PortableTextComponents, toPlainText } from "@portabletext/vue"
import { mount } from "@vue/test-utils";
import { h } from "vue";

import { nodeParse, transformToPortableText } from "../../../src"
import { PortableTextImage, PortableTextTable } from "../../../src/transformers/transformer-models";
import { resolveImage, resolveTable,toVueImageDefault } from '../../../src/utils/resolution/vue';

const dummyRichText: Elements.RichTextElement = {
  value: "<p>some text in a paragraph</p>",
  type: ElementType.RichText,
  images: [],
  linkedItemCodenames: [],
  linkedItems: [
    {
      system: {
        id: "99e17fe7-a215-400d-813a-dc3608ee0294",
        name: "test item",
        codename: "test_item",
        language: "default",
        type: "test",
        collection: "default",
        sitemapLocations: [],
        lastModified: "2022-10-11T11:27:25.4033512Z",
        workflowStep: "published",
        workflow: "default"
      },
      elements: {
        text_element: {
          type: ElementType.Text,
          name: "text element",
          value: "random text value",
        },
      },
    },
  ],
  links: [],
  name: "dummy",
};

const components: PortableTextComponents = {
  types: {
    image: ({ value }: PortableTextComponentProps<PortableTextImage>) =>
      resolveImage(value, h, toVueImageDefault),
    table: ({ value }: PortableTextComponentProps<PortableTextTable>) =>
      resolveTable(value, h, toPlainText),
  },
};

describe("PortableText Vue Renderer", () => {
  it("renders simple HTML from portable text", () => {
    const jsonTree = nodeParse(dummyRichText.value);
    const portableText = transformToPortableText(jsonTree);

    const wrapper = mount(PortableText, {
      props: {
        value: portableText,
      },
    });

    expect(wrapper.html()).toMatchInlineSnapshot(
      `"<p>some text in a paragraph</p>"`
    );
  });

  it("renders an image", () => {
    dummyRichText.value = `<p>some text before an asset</p><figure data-asset-id="bc6f3ce5-935d-4446-82d4-ce77436dd412" data-image-id="bc6f3ce5-935d-4446-82d4-ce77436dd412"><img src="https://assets-us-01.kc-usercontent.com:443/cec32064-07dd-00ff-2101-5bde13c9e30c/7d534724-edb8-4a6d-92f6-feb52be61d37/image1_w_metadata.jpg" data-asset-id="bc6f3ce5-935d-4446-82d4-ce77436dd412" data-image-id="bc6f3ce5-935d-4446-82d4-ce77436dd412" alt="alternative_text"></figure>`;
    const jsonTree = nodeParse(dummyRichText.value);
    const portableText = transformToPortableText(jsonTree);

    const wrapper = mount(PortableText, {
      props: {
        value: portableText,
        components: components,
      },
    });

    expect(wrapper.html()).toMatchInlineSnapshot(`
"<p>some text before an asset</p>
<img src="https://assets-us-01.kc-usercontent.com:443/cec32064-07dd-00ff-2101-5bde13c9e30c/7d534724-edb8-4a6d-92f6-feb52be61d37/image1_w_metadata.jpg" alt="alternative_text">"
`);
  });

  it("renders a table", () => {
    dummyRichText.value = "<table><tbody>\n  <tr><td>Ivan</td><td>Jiri</td></tr>\n  <tr><td>Ondra</td><td>Dan</td></tr>\n</tbody></table>";
    const jsonTree = nodeParse(dummyRichText.value);
    const portableText = transformToPortableText(jsonTree);

    const wrapper = mount(PortableText, {
      props: {
        value: portableText,
        components: components,
      },
    });

    expect(wrapper.html()).toMatchInlineSnapshot(`
"<table>
  <tr>
    <td>Ivan</td>
    <td>Jiri</td>
  </tr>
  <tr>
    <td>Ondra</td>
    <td>Dan</td>
  </tr>
</table>"
`);
  });
});
