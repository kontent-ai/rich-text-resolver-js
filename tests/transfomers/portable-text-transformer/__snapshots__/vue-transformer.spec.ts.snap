// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PortableText Vue Renderer renders a table 1`] = `
"<table>
  <tr>
    <td>Ivan</td>
    <td>Jiri</td>
  </tr>
  <tr>
    <td>Ondra</td>
    <td>Dan</td>
  </tr>
</table>"
`;

exports[`PortableText Vue Renderer renders an image 1`] = `
"<p>some text before an asset</p>
<img src="https://assets-us-01.kc-usercontent.com:443/cec32064-07dd-00ff-2101-5bde13c9e30c/7d534724-edb8-4a6d-92f6-feb52be61d37/image1_w_metadata.jpg" alt="alternative_text">"
`;

exports[`PortableText Vue Renderer renders simple HTML from portable text 1`] = `"<p>some text in a paragraph</p>"`;
